Template binary sensor attributes

gadgetchnnel opened the pull request:
## Description:

Added option to add custom attributes to the template binary sensor via a new attribute_templates config option. This is an updated version of PR #20410 concentrating on just the binary_sensor.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#9106

## Example entry for `configuration.yaml` (if applicable):
```yaml
binary_sensor:
  - platform: template
    sensors:
      test_sensor:
        value_template: >-
          {{ states.device_tracker.my_phone.state == "home" }}
        attribute_templates:
          latitude: >-
            {{ states.device_tracker.my_phone.attributes['latitude'] }}
          longitude: >-
            {{ states.device_tracker.my_phone.attributes['longitude'] }}
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

gadgetchnnel committed:
Added attribute support to template binary sensor with tests

Added attribute support to template binary sensor with tests

gadgetchnnel committed:
fix dictionary update

fix dictionary update

gadgetchnnel committed:
Fixed whitespace and line length issues

gadgetchnnel commented:
@rohankapoorcom Here is the new PR, concentrating on the binary_sensor.

gadgetchnnel committed:
Fixed indentation

rohankapoorcom commented:
I'll review this tonight.

pvizeli reviewed (commented):
Attributes are for the frontend. I.e. the GPS is on device tracker/person Entity defined. This feature ends up in some hacks between different layers. Otherside, some component implemented also a dynamic attribute handling.

I'm not a big fan but also don't want to dislike it. Maybe @balloob have another mind to this topic. That is only my personal one.

gadgetchnnel commented:
> Attributes are for the frontend. I.e. the GPS is on device tracker/person Entity defined. This feature ends up in some hacks between different layers. Otherside, some component implemented also a dynamic attribute handling.

There are other components which already allow attributes to be created, such as the RESTful sensor. Would a different example be better, one which doesn't just directly map attributes from other entities? Here is an actual example that I have used (with this code running as a custom component) which detects if my keys are lost (location more than 100 metres way) and has an attribute for the distance they are away in metres:

```
lost_keys:
  friendly_name: Keys Lost
  value_template: >-
    {{ (distance(states.device_tracker.keys, states.device_tracker.phone) * 1000) > 100 }}
  attribute_templates:
    distance: >-
      {{ distance(states.device_tracker.keys, states.device_tracker.phone) * 1000 }}
  entity_id:
    - sensor.keys
    - device_tracker.phone
```

Swamp-Ig commented:
This is good IMHO. I was planning something similar, but putting it in a base class for all the template and rest components so they all work the same. I have got that work ready to go, it's just waiting on getting some other PRs merged though, so don't roll this out to all the others yet.

Swamp-Ig commented:
Linking to PRs waiting to merge: #23590, #23293, Issue: #22587

Swamp-Ig commented:
Arch issue: home-assistant/architecture#220

MartinHjelmare commented:
What is the status and plan for this PR?

balloob commented on homeassistant/components/template/binary_sensor.py:
```diff
@@ -199,14 +217,27 @@ def _async_render(self):
+        ))
+
+        if self._attribute_templates is not None:
+            templates.update({'{}{}'.format(ATTRIBUTES_PREFIX, key): value
```
This is a very complicated, please simplify. Do not set attributes on the Entity class. Instead, this should be something like:

```python
attrs = {}
for key, value in self._attribute_templates.items():
    try:
        attrs[key] = value.async_render()
    catch TemplateError as err:
        _LOGGER.error("Error rendering attribute %s: %s, key, err)
        
self._attributes = attrs
```

gadgetchnnel replied:
I have now updated it to use that method (and added an additional test for this, since the extra exception handler pushed the coverage below the target).

balloob commented:
I like this approach.

gadgetchnnel committed:
Merge remote-tracking branch 'upstream/dev' into template-binary-sensor-attributes

gadgetchnnel committed:
Simplify applying of attribute templates based on feedback

gadgetchnnel committed:
Syntax and whitespace fixes

gadgetchnnel committed:
Black formatting

gadgetchnnel committed:
Black formatting on tests

gadgetchnnel committed:
Check attribute_templates is not None

gadgetchnnel committed:
Fixed test

gadgetchnnel committed:
Merge remote-tracking branch 'upstream/dev' into template-binary-sensor-attributes

gadgetchnnel committed:
Added test for failure to render template

gadgetchnnel committed:
Test fix

gadgetchnnel committed:
Updated test

gadgetchnnel committed:
Removed whitespace and applied Black formatting

gadgetchnnel committed:
Fixed test assertion

gadgetchnnel committed:
Updated test

damianromanowski commented:
Nice. Been wanting something like this.

balloob commented on homeassistant/components/template/binary_sensor.py:
```diff
@@ -59,14 +63,22 @@
+        )
+
+        if attribute_templates is not None:
+            templates.update(attribute_templates)
```
This is not good, it means attribute templates could override the other 3 templates.

balloob commented on homeassistant/components/template/binary_sensor.py:
```diff
@@ -59,14 +63,22 @@
-            (CONF_ICON_TEMPLATE, icon_template),
-            (CONF_ENTITY_PICTURE_TEMPLATE, entity_picture_template),
-        ):
+        templates = dict(
```
```python
templates = {
    CONF_VALUE_TEMPLATE: value_template,
    CONF_ICON_TEMPLATE, icon_template,
    # etc
}
```

balloob commented on homeassistant/components/template/binary_sensor.py:
```diff
@@ -59,14 +63,22 @@
+        if attribute_templates is not None:
+            templates.update(attribute_templates)
+
+        for tpl_name, template in templates.items():
```
Use `chain` to iterate over 2 iterables 

```python
from itertools import chain

for tpl_name, template in chain(templates.items(), attribute_templates.items()):
```

balloob commented on homeassistant/components/template/binary_sensor.py:
```diff
@@ -225,10 +246,23 @@ def _async_render(self):
-            ("_icon", self._icon_template),
-            ("_entity_picture", self._entity_picture_template),
-        ):
+        templates = dict(
```
Use `templates = { … }`

balloob commented on homeassistant/components/template/binary_sensor.py:
```diff
@@ -30,12 +30,16 @@
 CONF_DELAY_OFF = "delay_off"
+CONF_ATTRIBUTE_TEMPLATES = "attribute_templates"
+
+ATTRIBUTES_PREFIX = "_attributes."
```
unused

balloob reviewed (changes requested):
Left some comments.

gadgetchnnel committed:
Code improvements folloing comments

Using chain to iterate over templates and attribute_templates
Replacing dict() with {}
Rmoving unused constant

gadgetchnnel committed:
Applied Black formatting

gadgetchnnel committed:
Fixed removed code

gadgetchnnel committed:
Default attribute_templates to empty dict

gadgetchnnel committed:
Black formatting

gadgetchnnel committed:
Fixed imports

balloob commented:
Awesome  !  🐬

gadgetchnnel commented:
@balloob I also have a branch for attributes on the template sensor (I originally had both in the same branch but split them apart following earlier feedback).
I can submit a PR for this once I have checked that this includes all the changes from your feedback on this one.

OleksandrBerchenko commented:
@gadgetchnnel Hi, do you plan to create similar PRs for switch templates and light templates?  If not, I might do that myself.

Thanks in advance!
